Tests/MathLib/mechanics_invariants.cpp

```diff
@@ -0,0 +1,286 @@
+                  KelvinToTensorSize<6>::value> const& m)
+{
+    KelvinVector<6> v;
+    v << m(0, 0), m(1, 1), m(2, 2), m(0, 1) * std::sqrt(2.),
+        m(1, 2) * std::sqrt(2.), m(0, 2) * std::sqrt(2.);
+    return v;
+}
+
+struct MathLibMechanicsInvariants1 : public ::testing::Test
+{
+    static const int size = 1;
+    ac::randomEigenMatrixGenerator<double, size, 1> kelvinVectorGenerator;
+
+    ac::gtest_reporter gtest_reporter;
+};
+
+struct MathLibMechanicsInvariants4 : public ::testing::Test
+{
+    static const int size = 4;
+    ac::randomEigenMatrixGenerator<double, size, 1> kelvinVectorGenerator;
+
+    ac::gtest_reporter gtest_reporter;
+};
+
+struct MathLibMechanicsInvariants6 : public ::testing::Test
+{
+    static const int size = 6;
+    ac::randomEigenMatrixGenerator<double, size, 1> kelvinVectorGenerator;
+
+    ac::gtest_reporter gtest_reporter;
+};
+
+KelvinVector<4> setZcomponentZero(KelvinVector<4> const& v, std::size_t)
+{
+    auto result = v;
+    result[2] = 0;
+    return result;
+};
+
+//
+// Self-test of Kelvin to tensor and back conversion, which should be identity
+// up to numerical errors.
+//
+
+TEST_F(MathLibMechanicsInvariants1, SelfTestMappingKelvinToTensor)
+{
+    auto f = [](KelvinVector<1> const& v) {
+        return v == tensorToKelvin<1>(kelvinToTensor(v));
+    };
+
+    ac::check<KelvinVector<1>>(
+        f, 1000, ac::make_arbitrary(kelvinVectorGenerator), gtest_reporter);
+}
+
+TEST_F(MathLibMechanicsInvariants4, SelfTestMappingKelvinToTensor)
+{
+    auto f = [](KelvinVector<4> const& v) {
+        return (v - tensorToKelvin<4>(kelvinToTensor(v))).norm() <=
+               2 * std::numeric_limits<double>::epsilon() * v.norm();
+    };
+
+    ac::check<KelvinVector<4>>(
+        f, 1000,
+        ac::make_arbitrary(ac::map(&setZcomponentZero, kelvinVectorGenerator)),
+        gtest_reporter);
+}
+
+TEST_F(MathLibMechanicsInvariants6, SelfTestMappingKelvinToTensor)
+{
+    auto f = [](KelvinVector<6> const& v) {
+        return (v - tensorToKelvin<6>(kelvinToTensor(v))).norm() <=
+               1.5 * std::numeric_limits<double>::epsilon() * v.norm();
+    };
+
+    ac::check<KelvinVector<6>>(
+        f, 1000, ac::make_arbitrary(kelvinVectorGenerator), gtest_reporter);
+}
+
+//
+// Determinants of a Kelvin vector and corresponding Eigen::Matrix are equal.
+//
+
+TEST_F(MathLibMechanicsInvariants1, Determinant)
+{
+    auto f = [](KelvinVector<1> const& v) {
+        return Mechanics::Invariants<1>::determinant(v) ==
+               kelvinToTensor(v).determinant();
+    };
+
+    ac::check<KelvinVector<1>>(
+        f, 1000, ac::make_arbitrary(kelvinVectorGenerator), gtest_reporter);
+}
+
+TEST_F(MathLibMechanicsInvariants4, Determinant)
+{
+    auto f = [](KelvinVector<4> const& v) {
+        return std::abs(Mechanics::Invariants<4>::determinant(v) -
+                        kelvinToTensor(v).determinant()) <=
+               1e-12 * std::pow(v.norm(), 4);
```

Comment:
If ||x|| = 2000, then ||x||^4 = 16.0e+12, which means that the tolerance you apply is 16.0, which is not very strict. Or am I getting it wrong?
If the norms were bounded, e.g. 1/2 < ||x|| < 3/2 or ||x|| < 1, the applied tolerance as computed above (`1e-12*||x||^4`) would be bounded, too.